R/bmerge.R

```diff
@@ -25,6 +25,21 @@ coerce_col = function(dt, col, from_type, to_type, from_name, to_name, from_deta
   set(dt, j=col, value=cast_with_attrs(dt[[col]], cast_fun))
 }
 
+# data.table::as.ITime, chron::times, nanotime::nanotime
+.known_time_classes = c("Date", "POSIXt", "ITime", "times", "nanotime")
+.maybe_warn_mismatched_time_types = function(x_class, i_class, x_name, i_name) {
+  x_class_time = intersect(class(x[[xcol]]), .known_time_classes)
+  if (!length(x_class_time)) return(invisible())
+
+  i_class_time = intersect(class(i[[icol]]), .known_time_classes)
+  if (!length(i_class_time)) return(invisible())
+
+  if (identical(x_class_time, i_class_time)) return(invisible())
+
+  warningf("Attempting to join column %s (%s) with column %s (%s). They are likely to be incompatible, so we suggest you convert one to the other's class.",
```

Comment:
"they are likely to be incompatible" feels like not the best wording. What's meant by "incompatible" here?

One thing I have in mind is, it'd be nice to recommend the user which conversion to do, e.g. they should only do POSIXct-->Date if they're OK with some precision loss, whereas they can do Date-->POSIXct WLOG.

OTOH, nanotime and POSIXt may be incompatible in a much deeper sense, that there are times mutually non-representable between 64-bit storage epoch nanosecond precision and 64-bit epoch second precision.

Reply:
I don't disagree, though while it's always nice to suggest a remedy, with several less-unambiguous class-comparisons I didn't spend the time to think through all of them. Perhaps that was me being a little lazy at the time.